lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp

```diff
@@ -229,32 +224,54 @@ bool ProxyRenderDelegate::requiresUpdate(const MSubSceneContainer& container, co
     return true;
 }
 
+void ProxyRenderDelegate::_ClearRenderDelegate()
+{
+    // The order of deletion matters. Some orders cause crashes.
+    // This order matches the deletion order when a ProxyRenderDelegate is destroyed.
+    // Class member variables are destroyed in the reverse order in which they appear in the class declaration.
+
+    _sceneDelegate.reset(nullptr);
+    _taskController.reset(nullptr);
+    _renderIndex.reset(nullptr);
+    _renderDelegate.reset(nullptr);
+}
+
 //! \brief  One time initialization of this drawing routine
-void ProxyRenderDelegate::_InitRenderDelegate() {
-    // No need to run all the checks if we got till the end
-    if (_isInitialized())
-        return;
+void ProxyRenderDelegate::_InitRenderDelegate(MSubSceneContainer& container) {
 
     if (_proxyShape == nullptr)
         return;
 
-    if (!_usdStage) {
-        _usdStage = _proxyShape->getUsdStage();
+    UsdStageRefPtr newUsdStage = _proxyShape->getUsdStage();
+    if (_usdStage != newUsdStage)
```

Comment:
I was thinking more about this. We currently have two code which needs to re-populate the render index:
- file path change
- exclude prim paths change

If you look at the code in _Populate method, you will notice `_proxyShape->getExcludePrimPathsVersion() != _excludePrimPathsVersion`. I believe this code no longer works as expected and we will have to do a very similar thing when excluded prims change. 

How about we merge the two together and instead of testing the stage at line 246, you will test `stageVersion`? Then in proxyShapeBase, we would bump the stageVersion on excluded prims change (basically rename _IncreaseExcludePrimPathsVersion to _IncreaseStageVersion) and from https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/nodes/proxyShapeBase.cpp#L531

Reply:
Yes I think it does make sense to combine those two ideas. Is the MayaUsdProxyShapeBase a class that AL and Pixar derive from but I don't see them making any calls to getExcludePrimPathsVersion(), so we should be good to change this functionality a little bit without much impact on them.

In the future if there is a scenario where we'd want more fine grained information about what changed (the exclude prim paths vs. the stage) we could split the version number into two, one for exclude paths and one for the stage. Right now we'd be doing the same work either way so combining them should work.

Reply:
I went in a started making the changes and I learned a few new things.

1. excludePrimPathsAttr doesn't affect outStageDataAttr, so combing changes to excludePrimPathsAttr with changes to outStageDataAttr into a single version counter implies some knowledge of the consumers of that data. In our case we know that either one changing will require us to call populate again, but MayaUsdProxyShapeBase doesn't know that.

1. There _are_ some other attributes that do affect what ends up in outStageDataAttr: filePathAttr as we already identified, but loadPayloadsAttr also impacts it. And kind of indirectly primPathAttr.

1. primPathAttr is a little different because the data in outStageDataAttr is a MayaUsdStageData which has two values, the UsdStage & the SdfPath prim path.

I think in MayaUsdProxyShapeBase it is wrong to combine dirty information because that implies knowledge of the downstream consumers of that information, and we don't control all the downstream consumers. That means that for example the dirty counter _ExcludePrimPathVersion is fine, but a combined counter for _IncreaseStageVersion that increments when excludePrimPathsAttr changes is not fine.

We need separate version counters for all the output data on MayaUsdProxyShapeBase. Those counters can be on the proxy shape itself, or they can be in the render delegates (we'd have to create a mixin class with the functionality so customers could re-use it). I think it's probably simpler to just put it on the ProxyShapeBase.

We could also put the version counters for the stage and primPath inside the MayaUsdStageData.

What I will do is for each output attribute create the version counter on MayaUsdProxyShapeBase and set them up to increment properly. That way everyone can use them. Then in ProxyRenderDelegate we'll have to track all the version ids we care about to do the updates when necessary. Once we have that basic version in the PR we can decide what tweaks we want to the system.

Reply:
We don't need version counters for simple numeric types which are easy to store & compare vs. their old value.

Reply:
The new version is going the direction I think we should take.